edb/schema/functions.py

```diff
@@ -291,20 +298,28 @@ def get_verbosename(
         else:
             return vn
 
-    def get_shortname(self, schema: s_schema.Schema) -> str:
+    # type ignore below, because `get_shortname` in objects.QualifiedObject
+    # has return type of sn.Name; the following method instead *needs* to
+    # return a str
+    def get_shortname(self, schema: s_schema.Schema) -> str:  # type: ignore
```

Comment:
This doesn't seem to be sound. Can you try removing this implementation and use the default one instead?

Reply:
I looked into this for a while, and I would like your guidance to know the approach you prefer.
The `get_shortname` from the ancestor `QualifiedObject` is annotated and designed to return a `sn.SchemaName`: all callers are using exactly this type and wouldn't work with a regular `str`.
The `get_shortname` from the "interface" `ParameterLike` is annotated to return a `str`. The `Parameter` class is a descendant of both classes, and it implements the signature defined in `ParameterLike`, in fact a `sn.Name` is unwrapped into a `str` inside the method `paramname_from_fullname`.

There are currently 7 annotated functions `get_shortname`: 5 of them are annotated to return `str`, and 2 of them to return a more specialised `SchemaName` (QualifiedObject and EphemeralCollection). Both the functions in `QualifiedObject` and `Parameter` class are used by callers that rely on `SchemaName` and `str` respectively.

To make the functions consistent, we can for example refactor the code in `ParameterLike` to return `SchemaName`. In such case, how the following function should be modified? Would it make sense to return a SchemaName having `module` property equal to `fullname` input parameter?

```python
    @classmethod
    def paramname_from_fullname(cls, fullname: sn.Name) -> str:
        parts = str(fullname.name).split('@@', 1)
        if len(parts) == 2:
            return sn.unmangle_name(parts[0])
        elif '::' in fullname:
            return sn.Name(fullname).name
        else:
            return fullname
```

Reply:
The callers that rely on `Parameter.get_shortname` returning `str` should call another function that wraps `get_shortname`, and `Parameter.get_shortname` should be removed (i.e. revert to the implementation in `QualifiedObject`):

```python
class Parameter:
    ...
    def get_parameter_name(self, schema) -> str:
        return self.get_shortname(schema).name
```

Reply:
Thanks Elvis for the instructions - I was in doubt earlier that you indeed meant this.
I worked in that direction, you can see in [this commit](https://github.com/edgedb/edgedb/commit/00af9ca4b1a32737b3870b941955aa4dc4b54a56) the files I had to change to be able to run the tests (to get the test database populated properly). Several tests are failing, and I get several of these:

```
    await self._protocol.simple_query(query)
  File "edgedb/protocol/protocol.pyx", line 497, in simple_query
edgedb.errors.InvalidReferenceError: improperly formed name: module is not specified: a
```

To find out when the old function is called, I added a function that logs the trace whenever the `Parameter.get_shortname` function is used; and it seems the only place where it is still used is from `/edgedb/edb/schema/objects.py", line 656, in get_displayname
    return str(self.get_shortname(schema))` - but of course I don't want to change this piece of code that is rightly abstracted from `Parameter` class to do something else. But maybe I only need to modify still the `get_verbosename` in Functions. Just writing this is giving me a bit more of insight. :smile:

Reply:
Right. `Parameter` name generation is a bit backwards.  Here's how I propose to fix this:

```python
    def get_shortname(self, schema: s_schema.Schema) -> sn.Name:
        return sn.Name(
            module=self.get_name(schema).module,
            name=self.get_parameter_name(schema),
        )

    def get_displayname(self, schema: s_schema.Schema) -> str:
        return self.get_parameter_name(schema)
```

Reply:
It looks good now!